crates/services/p2p/src/cached_view.rs

```diff
@@ -0,0 +1,320 @@
+use crate::ports::P2pDb;
+use dashmap::DashMap;
+use fuel_core_metrics::p2p_metrics::{
+    increment_p2p_req_res_cache_hits,
+    increment_p2p_req_res_cache_misses,
+};
+use fuel_core_storage::Result as StorageResult;
+use fuel_core_types::{
+    blockchain::SealedBlockHeader,
+    services::p2p::Transactions,
+};
+use std::ops::Range;
+type BlockHeight = u32;
+
+pub(super) struct CachedView {
+    sealed_block_headers: DashMap<BlockHeight, SealedBlockHeader>,
+    transactions_on_blocks: DashMap<BlockHeight, Transactions>,
+    metrics: bool,
+}
+
+impl CachedView {
+    pub fn new(metrics: bool) -> Self {
+        Self {
+            sealed_block_headers: DashMap::new(),
+            transactions_on_blocks: DashMap::new(),
+            metrics,
+        }
+    }
+
+    pub(super) fn clear(&self) {
+        self.sealed_block_headers.clear();
+        self.transactions_on_blocks.clear();
+    }
+
+    fn update_metrics<U>(&self, update_fn: U)
+    where
+        U: FnOnce(),
+    {
+        if self.metrics {
+            update_fn()
+        }
+    }
+
+    fn get_from_cache_or_db<V, T, F>(
+        &self,
+        cache: &DashMap<u32, T>,
+        view: &V,
+        range: Range<u32>,
+        fetch_fn: F,
+    ) -> StorageResult<Option<Vec<T>>>
+    where
+        V: P2pDb,
+        T: Clone,
+        F: Fn(&V, Range<u32>) -> StorageResult<Option<Vec<T>>>,
+    {
+        let mut items = Vec::new();
+        let mut missing_start = None;
+
+        for height in range.clone() {
+            if let Some(item) = cache.get(&height) {
+                items.push(item.clone());
```

Comment:
It can be follow up PR, but it would be nice if we avoid heavy clone here and used `Arc` instead

Reply:
added comment here - d897cba69c

Reply:
associated issue: https://github.com/FuelLabs/fuel-core/issues/2436